File: src/BlazorAjaxToolkitComponents/PopupControlExtender.cs
using BlazorAjaxToolkitComponents.Enums;
using Microsoft.AspNetCore.Components;

namespace BlazorAjaxToolkitComponents;

/// <summary>
/// Attaches a popup panel to a target control, displaying it on click.
/// Lighter than ModalPopupExtender — no overlay, no focus trap.
/// Supports positional placement, commit property/script, and outside-click dismissal.
/// Emulates the Ajax Control Toolkit PopupControlExtender.
/// </summary>
public class PopupControlExtender : BaseExtenderComponent
{
	private const string MODULE_PATH =
		"./_content/BlazorAjaxToolkitComponents/js/popup-control-extender.js";

	/// <summary>
	/// The ID of the panel element to display as a popup.
	/// </summary>
	[Parameter]
	public string PopupControlID { get; set; } = string.Empty;

	/// <summary>
	/// The position of the popup relative to the target control.
	/// </summary>
	[Parameter]
	public PopupPosition Position { get; set; } = PopupPosition.Bottom;

	/// <summary>
	/// Horizontal offset in pixels from the calculated position.
	/// </summary>
	[Parameter]
	public int OffsetX { get; set; }

	/// <summary>
	/// Vertical offset in pixels from the calculated position.
	/// </summary>
	[Parameter]
	public int OffsetY { get; set; }

	/// <summary>
	/// The property name on the target control to set when the popup is committed.
	/// Typically "value" for input elements.
	/// </summary>
	[Parameter]
	public string CommitProperty { get; set; } = string.Empty;

	/// <summary>
	/// JavaScript to execute when the popup is committed (a value is selected).
	/// </summary>
	[Parameter]
	public string CommitScript { get; set; } = string.Empty;

	/// <summary>
	/// The ID of this extender control for programmatic access.
	/// </summary>
	[Parameter]
	public string ExtenderControlID { get; set; } = string.Empty;

	protected override string JsModulePath => MODULE_PATH;

	protected override string JsCreateFunction => "createBehavior";

	protected override object GetBehaviorProperties() => new
	{
		popupControlId = PopupControlID,
		position = (int)Position,
		offsetX = OffsetX,
		offsetY = OffsetY,
		commitProperty = CommitProperty,
		commitScript = CommitScript,
		extenderControlId = ExtenderControlID
	};
}

File: src/BlazorAjaxToolkitComponents/wwwroot/js/popup-control-extender.js
// PopupControlExtender JS behavior module
// Displays a popup panel attached to a target control on click.
// Lighter than ModalPopupExtender — no overlay, no focus trap.
// Supports positional placement, commit property/script, and outside-click dismiss.

const behaviors = new Map();

// PopupPosition (matches C# enum)
const POSITION_LEFT = 0;
const POSITION_RIGHT = 1;
const POSITION_TOP = 2;
const POSITION_BOTTOM = 3;
const POSITION_CENTER = 4;

/**
 * Creates a popup control behavior and attaches it to the target element.
 * @param {{ targetId: string, behaviorId: string, properties: object }} config
 * @returns {object} A JS object reference for the behavior instance.
 */
export function createBehavior(config) {
    const { targetId, behaviorId, properties } = config;
    const trigger = document.getElementById(targetId);

    if (!trigger) {
        console.warn(`[PopupControlExtender] Target element '${targetId}' not found.`);
        return {};
    }

    const popup = document.getElementById(properties.popupControlId);
    if (!popup) {
        console.warn(`[PopupControlExtender] Popup element '${properties.popupControlId}' not found.`);
        return {};
    }

    const state = {
        targetId,
        properties: { ...properties },
        popup,
        isOpen: false,
        handlers: {}
    };

    // Hide popup initially
    popup.style.display = "none";

    function positionPopup() {
        const triggerRect = trigger.getBoundingClientRect();
        const popupRect = popup.getBoundingClientRect();
        const scrollX = window.scrollX || window.pageXOffset;
        const scrollY = window.scrollY || window.pageYOffset;
        const offsetX = properties.offsetX || 0;
        const offsetY = properties.offsetY || 0;

        let left, top;

        switch (properties.position) {
            case POSITION_LEFT:
                left = triggerRect.left - popupRect.width + scrollX + offsetX;
                top = triggerRect.top + scrollY + offsetY;
                break;
            case POSITION_RIGHT:
                left = triggerRect.right + scrollX + offsetX;
                top = triggerRect.top + scrollY + offsetY;
                break;
            case POSITION_TOP:
                left = triggerRect.left + scrollX + offsetX;
                top = triggerRect.top - popupRect.height + scrollY + offsetY;
                break;
            case POSITION_CENTER:
                left = triggerRect.left + (triggerRect.width - popupRect.width) / 2 + scrollX + offsetX;
                top = triggerRect.top + (triggerRect.height - popupRect.height) / 2 + scrollY + offsetY;
                break;
            case POSITION_BOTTOM:
            default:
                left = triggerRect.left + scrollX + offsetX;
                top = triggerRect.bottom + scrollY + offsetY;
                break;
        }

        popup.style.left = `${Math.max(0, left)}px`;
        popup.style.top = `${Math.max(0, top)}px`;
    }

    function show() {
        if (state.isOpen) return;
        state.isOpen = true;

        popup.style.display = "";
        popup.style.position = "absolute";
        popup.style.zIndex = "9999";
        positionPopup();

        // Add outside-click dismiss with a short delay to avoid immediate dismissal
        requestAnimationFrame(() => {
            document.addEventListener("mousedown", state.handlers.outsideClick);
        });
    }

    function hide() {
        if (!state.isOpen) return;
        state.isOpen = false;

        popup.style.display = "none";
        document.removeEventListener("mousedown", state.handlers.outsideClick);
    }

    function commit(value) {
        // Set commit property on target
        if (properties.commitProperty && value !== undefined) {
            trigger[properties.commitProperty] = value;
        }

        // Run commit script
        if (properties.commitScript) {
            try {
                new Function(properties.commitScript)();
            } catch (e) {
                console.error("[PopupControlExtender] Commit script error:", e);
            }
        }

        hide();
    }

    // Outside click handler
    state.handlers.outsideClick = function (e) {
        if (!popup.contains(e.target) && !trigger.contains(e.target)) {
            hide();
        }
    };

    // Trigger click toggles the popup
    state.handlers.triggerClick = function (e) {
        e.preventDefault();
        if (state.isOpen) {
            hide();
        } else {
            show();
        }
    };
    trigger.addEventListener("click", state.handlers.triggerClick);

    // Escape key closes popup
    state.handlers.keydown = function (e) {
        if (e.key === "Escape" && state.isOpen) {
            hide();
        }
    };
    document.addEventListener("keydown", state.handlers.keydown);

    state.show = show;
    state.hide = hide;
    state.commit = commit;
    behaviors.set(behaviorId, state);
    return {};
}

/**
 * Updates behavior properties.
 * @param {string} behaviorId
 * @param {object} properties
 */
export function updateBehavior(behaviorId, properties) {
    const state = behaviors.get(behaviorId);
    if (!state) return;
    state.properties = { ...properties };
}

/**
 * Disposes the behavior and removes event listeners.
 * @param {string} behaviorId
 */
export function disposeBehavior(behaviorId) {
    const state = behaviors.get(behaviorId);
    if (!state) return;

    // Close if open
    if (state.isOpen) {
        state.hide();
    }

    // Remove trigger listener
    const trigger = document.getElementById(state.targetId);
    if (trigger && state.handlers.triggerClick) {
        trigger.removeEventListener("click", state.handlers.triggerClick);
    }

    document.removeEventListener("keydown", state.handlers.keydown);
    document.removeEventListener("mousedown", state.handlers.outsideClick);

    behaviors.delete(behaviorId);
}
